mapreduce/mapreduce-base/src/main/java/io/tidb/bigdata/mapreduce/tidb/TiDBInputFormat.java

```diff
@@ -0,0 +1,206 @@
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+/**
+ * A InputFormat that reads input data from an TiDB table.
+ * <p>
+ * DBInputFormat emits LongWritables containing the record number as
+ * key and TiDBWritables as value.
+ * <p>
+ */
+public class TiDBInputFormat<T extends TiDBWritable>
+    extends InputFormat<LongWritable, T> implements Configurable {
+
+  private List<ColumnHandleInternal> columnHandleInternals;
+
+  private ClientSession clientSession;
+
+  private TableHandleInternal tableHandleInternal;
+
+  private TiDBConfiguration dbConf;
+
+  private ResultSetMetaData resultSetMetaData;
+
+  public TiDBInputFormat() {
+
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @return
+   */
+  @Override
+  public List<InputSplit> getSplits(JobContext job) {
+    List<InputSplit> splits = new ArrayList<>();
+    SplitManagerInternal splitManagerInternal = new SplitManagerInternal(clientSession);
+    List<SplitInternal> splitInternals = splitManagerInternal.getSplits(tableHandleInternal);
+    for (SplitInternal splitInternal : splitInternals) {
+      splits.add(new TiDBInputSplit(splitInternal.getStartKey(), splitInternal.getEndKey(),
+          tableHandleInternal.getConnectorId(), tableHandleInternal.getSchemaName(),
+          tableHandleInternal.getTableName()));
+    }
+    return splits;
+  }
+
+  @Override
+  public RecordReader<LongWritable, T> createRecordReader(InputSplit inputSplit,
+      TaskAttemptContext taskAttemptContext) {
+
+    return new TiDBRecorderReader((TiDBInputSplit) inputSplit, getConf(), getClientSession(),
+        columnHandleInternals, resultSetMetaData);
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    this.dbConf = new TiDBConfiguration(conf);
+    getClientSession();
+
+    String databaseName = dbConf.getDatabaseName();
+
+    String tableName = dbConf.getInputTableName();
+
+    // check database and table
+    clientSession.getTableMust(databaseName, tableName);
+
+    this.tableHandleInternal = new TableHandleInternal(
+        UUID.randomUUID().toString(), databaseName, tableName);
+    List<ColumnHandleInternal> columns = clientSession.getTableColumns(tableHandleInternal)
+        .orElseThrow(() -> new NullPointerException("columnHandleInternals is null"));
+    Map<String, Integer> nameAndIndex = new HashMap<>();
+
+    IntStream.range(0, columns.size())
+        .forEach(i -> nameAndIndex.put(columns.get(i).getName(), i));
+
+    String[] fieldNames =
```

Comment:
what if `fieldNames.length = 0` or ` fieldNames = null` ?

Reply:
it‘s going to be judged by the setInput method in class TiDBInputFormat.